File: web/src/domain/progress.ts
import { roundToTenth } from '@/domain/formulas';

export const DAY_MS = 86_400_000;

export type ProgressMetric = 'maxWeightKg' | 'volumeKg' | 'bestOneRepMaxKg';

export type RangePreset = '30d' | '90d' | 'all';

export interface ProgressRange {
  readonly fromMs: number | null;
  readonly toMs: number;
}

export interface ProgressSetInput {
  readonly sessionId: string;
  readonly startedAtMs: number;
  readonly weightKg: number | null;
  readonly reps: number | null;
  readonly isWarmup: boolean;
}

export interface ProgressPoint {
  readonly sessionId: string;
  readonly startedAtMs: number;
  readonly maxWeightKg: number;
  readonly volumeKg: number;
  readonly bestOneRepMaxKg: number;
  readonly workingSets: number;
}

export function rangeFor(preset: RangePreset, nowMs: number): ProgressRange | null {
  if (preset === 'all') {
    return null;
  }
  const days = preset === '30d' ? 30 : 90;
  return { fromMs: nowMs - days * DAY_MS, toMs: nowMs };
}

export function buildProgressSeries(
  sets: readonly ProgressSetInput[],
  range: ProgressRange | null = null
): ProgressPoint[] {
  const buckets = new Map<string, ProgressPoint>();

  for (const set of sets) {
    if (set.isWarmup) {
      continue;
    }
    if (range !== null) {
      if (range.fromMs !== null && set.startedAtMs < range.fromMs) {
        continue;
      }
      if (set.startedAtMs > range.toMs) {
        continue;
      }
    }

    const existing = buckets.get(set.sessionId);
    const workingSets = (existing?.workingSets ?? 0) + 1;
    const hasValues = set.weightKg !== null && set.reps !== null;

    const maxWeightKg = hasValues
      ? Math.max(existing?.maxWeightKg ?? 0, set.weightKg as number)
      : existing?.maxWeightKg ?? 0;

    const volumeKg = hasValues
      ? (existing?.volumeKg ?? 0) + (set.weightKg as number) * (set.reps as number)
      : existing?.volumeKg ?? 0;

    const bestOneRepMaxKg = hasValues
      ? Math.max(
          existing?.bestOneRepMaxKg ?? 0,
          roundToTenth((set.weightKg as number) * (1 + (set.reps as number) / 30))
        )
      : existing?.bestOneRepMaxKg ?? 0;

    buckets.set(set.sessionId, {
      sessionId: set.sessionId,
      startedAtMs: existing?.startedAtMs ?? set.startedAtMs,
      maxWeightKg,
      volumeKg,
      bestOneRepMaxKg,
      workingSets,
    });
  }

  return [...buckets.values()].sort((a, b) => {
    if (a.startedAtMs !== b.startedAtMs) {
      return a.startedAtMs - b.startedAtMs;
    }
    return a.sessionId < b.sessionId ? -1 : a.sessionId > b.sessionId ? 1 : 0;
  });
}

export function metricValue(point: ProgressPoint, metric: ProgressMetric): number {
  return point[metric];
}

export function formatMetric(value: number, metric: ProgressMetric): string {
  if (metric === 'volumeKg') {
    return `${Math.round(value).toLocaleString('es')} kg`;
  }
  return `${Math.round(value * 10) / 10} kg`;
}

export function formatDay(timestampMs: number): string {
  return new Date(timestampMs).toLocaleDateString('es', { day: '2-digit', month: '2-digit' });
}

File: android/app/src/main/java/com/fitlog/app/domain/Progress.kt
package com.fitlog.app.domain

object Progress {

    const val DAY_MS = 86_400_000L

    enum class Metric { MAX_WEIGHT, VOLUME, BEST_ONE_REP_MAX }

    enum class RangePreset { LAST_30_DAYS, LAST_90_DAYS, ALL }

    data class Range(val fromMs: Long?, val toMs: Long)

    data class SetInput(
        val sessionId: String,
        val startedAtMs: Long,
        val weightKg: Double?,
        val reps: Int?,
        val isWarmup: Boolean,
    )

    data class Point(
        val sessionId: String,
        val startedAtMs: Long,
        val maxWeightKg: Double,
        val volumeKg: Double,
        val bestOneRepMaxKg: Double,
        val workingSets: Int,
    )

    fun rangeFor(preset: RangePreset, nowMs: Long): Range? = when (preset) {
        RangePreset.ALL -> null
        RangePreset.LAST_30_DAYS -> Range(fromMs = nowMs - 30 * DAY_MS, toMs = nowMs)
        RangePreset.LAST_90_DAYS -> Range(fromMs = nowMs - 90 * DAY_MS, toMs = nowMs)
    }

    fun buildSeries(sets: List<SetInput>, range: Range? = null): List<Point> {
        val buckets = linkedMapOf<String, Point>()

        for (set in sets) {
            if (set.isWarmup) continue
            if (range != null) {
                val from = range.fromMs
                if (from != null && set.startedAtMs < from) continue
                if (set.startedAtMs > range.toMs) continue
            }

            val existing = buckets[set.sessionId]
            val workingSets = (existing?.workingSets ?: 0) + 1
            val weight = set.weightKg
            val reps = set.reps
            val complete = weight != null && reps != null

            val maxWeightKg = if (complete) {
                maxOf(existing?.maxWeightKg ?: 0.0, weight)
            } else {
                existing?.maxWeightKg ?: 0.0
            }

            val volumeKg = if (complete) {
                (existing?.volumeKg ?: 0.0) + weight * reps
            } else {
                existing?.volumeKg ?: 0.0
            }

            val bestOneRepMaxKg = if (complete) {
                maxOf(
                    existing?.bestOneRepMaxKg ?: 0.0,
                    Formulas.roundToTenth(weight * (1 + reps / 30.0)),
                )
            } else {
                existing?.bestOneRepMaxKg ?: 0.0
            }

            buckets[set.sessionId] = Point(
                sessionId = set.sessionId,
                startedAtMs = existing?.startedAtMs ?: set.startedAtMs,
                maxWeightKg = maxWeightKg,
                volumeKg = volumeKg,
                bestOneRepMaxKg = bestOneRepMaxKg,
                workingSets = workingSets,
            )
        }

        return buckets.values.sortedWith(
            compareBy({ it.startedAtMs }, { it.sessionId })
        )
    }

    fun metricValue(point: Point, metric: Metric): Double = when (metric) {
        Metric.MAX_WEIGHT -> point.maxWeightKg
        Metric.VOLUME -> point.volumeKg
        Metric.BEST_ONE_REP_MAX -> point.bestOneRepMaxKg
    }
}
